File: Python3/1001 - 1200/1026 Maximum Difference Between Node and Ancestor.py
# 40ms, 75.03%; 17.5MB, 88.89%
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxAncestorDiff(self, root: TreeNode) -> int:
        self.ret = 0
        self.__helper(root)
        return self.ret

    def __helper(self, root: TreeNode) -> (int, int):
        if root.left is root.right: return (root.val, root.val)
        _min, _max = 100000, 0
        if root.left:
            _min, _max = self.__helper(root.left)
        if root.right:
            x, y = self.__helper(root.right)
            _min, _max = min(_min, x), max(_max, y)
        self.ret = max(self.ret, abs(root.val - _min), abs(root.val - _max))
        return (min(_min, root.val), max(_max, root.val))

File: 1001 - 1200/1026 Maximum Difference Between Node and Ancestor.cc
// 4ms, 97.12%; 15.4MB, 18.18%
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    int maxAncestorDiff(TreeNode *root) {
        int ret = 0;
        helper(root, ret);
        return ret;
    }

private:
    pair<int, int> helper(TreeNode *node, int &ret) {
        if (node->left == node->right) return make_pair(node->val, node->val);
        int _max = 0, _min = 100000;
        if (node->left) {
            auto p1 = helper(node->left, ret);
            _max = max(p1.first, p1.second);
            _min = min(p1.first, p1.second);
        }
        if (node->right) {
            auto p2 = helper(node->right, ret);
            _max = max({_max, p2.first, p2.second});
            _min = min({_min, p2.first, p2.second});
        }
        ret = max({ret, abs(node->val - _max), abs(node->val - _min)});
        return make_pair(min(node->val, _min), max(node->val, _max));
    }
};
